ddtrace/settings.py

```diff
@@ -115,5 +117,133 @@ def __repr__(self):
+
+            @config.falcon.hooks.register('request')
+            def on_request(span, request, response):
+                pass
+
+        :param key: Hook name to register
+        :type key: str
+        :returns: A function decorator used to register a hook
+        :rtype: function
+        """
+        return self.register(key)
+
+    def register(self, hook, func=None):
+        """
+        Function used to register a hook for the provided name.
+
+        Example::
+
+            def on_request(span, request, response):
+                pass
+
+            config.falcon.hooks.register('request', on_request)
+
+
+        If no function is provided then a decorator is returned::
+
+            @config.falcon.hooks.register('request')
+            def on_request(span, request, response):
+                pass
+
+        :param hook: The name of the hook to register the function for
+        :type hook: str
+        :param func: The function to register, or ``None`` if a decorator should be returned
+        :type func: function, None
+        :returns: Either a function decorator if ``func is None``, otherwise ``None``
+        :rtype: function, None
+        """
+        # If they didn't provide a function, then return a decorator
+        if not func:
+            def wrapper(func):
+                self.register(hook, func)
+                return func
+            return wrapper
+        self._hooks[hook].add(func)
+
+    def deregister(self, func):
+        """
+        Function to deregister a function from all hooks it was registered under
+
+        Example::
+
+            @config.falcon.hooks.request
+            def on_request(span, request, response):
+                pass
+
+            config.falcon.hooks.deregister(on_request)
+
+
+        :param func: Function hook to register
+        :type func: function
+        """
+        for funcs in self._hooks.values():
+            if func in funcs:
+                funcs.remove(func)
+
+    def _emit(self, hook, span, *args, **kwargs):
+        """
+        Function used to call registered hook functions.
+
+        :param hook: The hook to call functions for
+        :type hook: str
+        :param span: The span to call the hook with
+        :type span: :class:`ddtrace.span.Span`
+        :param *args: Positional arguments to pass to the hook functions
+        :type args: list
+        :param **kwargs: Keyword arguments to pass to the hook functions
+        :type kwargs: dict
+        """
+        # Return early if no hooks are registered
+        if hook not in self._hooks:
+            return
+
+        # Return early if we don't have a Span
+        if not isinstance(span, Span):
+            return
+
+        # Call registered hooks
+        for func in self._hooks[hook]:
+            try:
+                func(span, *args, **kwargs)
+            except Exception as e:
+                log.debug('Failed to run hook {} function {}: {}'.format(hook, func, e))
```

Comment:
Shouldn't it be `log.error`?

Reply:
I'll add a comment but no, basically because we use the built in logger we don't have a way to throttle any logging we do.

```suggestion
                # DEV: Use log.debug instead of log.error until we have a throttled logger
                log.debug('Failed to run hook {} function {}: {}'.format(hook, func, e))
```